lsp/nls/src/linearization/building.rs

```diff
@@ -11,58 +12,78 @@ use nickel_lang::{
-    Environment, IdGen, LinearizationItem,
+    Environment, IdGen, ItemID, LinearizationItem,
 };
 
 /// A concrete [LinearizationState]
 /// Holds any inner datatype that can be used as stable resource
 /// while recording terms.
-#[derive(Default)]
-pub struct Building {
+// #[derive(Default)]
+pub struct Building<'a> {
     pub linearization: Vec<LinearizationItem<Unresolved>>,
+    pub lin_cache: &'a mut HashMap<FileId, Completed>,
 }
 
-pub type ID = usize;
-
-impl Building {
+impl<'b> Building<'b> {
     pub(super) fn push(&mut self, item: LinearizationItem<Unresolved>) {
         self.linearization.push(item)
     }
 
-    pub(super) fn add_usage(&mut self, decl: usize, usage: usize) {
-        match self
-            .linearization
-            .get_mut(decl)
-            .expect("Could not find parent")
-            .kind
-        {
-            TermKind::Structure => unreachable!(),
-            TermKind::Usage(_) => unreachable!(),
-            TermKind::Record(_) => unreachable!(),
+    pub(super) fn add_usage(&mut self, file: FileId, decl: ItemID, usage: ItemID) {
+        let (decl_file, index) = decl;
+        let kind = if file == decl_file {
+            // This usage references an item in the file we're currently linearizing
+            &mut self
+                .linearization
+                .get_mut(index)
+                .expect("Could not find parent")
+                .kind
+        } else {
+            // This usage references an item in another file (that has already been linearized)
+            &mut self
+                .lin_cache
+                .get_mut(&decl_file)
+                .expect("Could not find parent")
+                .get_item_mut(decl)
+                .expect("Could not find parent")
+                .kind
+        };
+
+        match kind {
+            TermKind::Record(_) | TermKind::Structure | TermKind::Usage(_) => unreachable!(),
             TermKind::Declaration(_, ref mut usages, _)
             | TermKind::RecordField { ref mut usages, .. } => usages.push(usage),
         };
     }
 
-    pub(super) fn inform_declaration(&mut self, declaration: ID, value: ID) {
-        if let Some(LinearizationItem {
-            kind: TermKind::Declaration(_, _, value_state),
-            ..
-        }) = self.linearization.get_mut(declaration)
-        {
+    pub(super) fn inform_declaration(&mut self, file: FileId, declaration: ItemID, value: ItemID) {
+        let (decl_file, index) = declaration;
+        let kind = if file == decl_file {
+            // This usage references an item in the file we're currently linearizing
+            self.linearization.get_mut(index).map(|item| &mut item.kind)
+        } else {
+            // This usage references an item in another file (that has already been linearized)
+            self.lin_cache
+                .get_mut(&decl_file)
+                .and_then(|lin| lin.get_item_mut(declaration).map(|item| &mut item.kind))
+        };
```

Comment:
It's the second time that this snippet appears: branching on `if file == decl_file` to get a mutable reference to an item. Maybe this could be put in a separate function in building, something like `get_item_mut(&mut self, file: FileId, item: ItemID)`.